torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1099,6 +1117,286 @@ def call_function(
+    of stuff associated with tracing for loops. That itself provides some speedup.
+
+    Once we implement function compilation caching, or a higher order operator,
+    like jax.lax.fori_loop, for loops will compile in constant time, and perhaps even
+    better than what we currently have.
+    """
+
+    class UnitializedVariable(ConstantVariable):
+        """Special sentinel to indicate values are not yet set. This differs
+        from misc.UnknownVariable in that it's actually known outside of a for loop
+        higher order op.
+        """
+
+        pass
+
+    # Any of these nested means the control flow is too complex to be
+    # represented by a higher order op.
+    FORBIDDEN_OPCODES = {
+        "RETURN_VALUE",
+        "YIELD_VALUE",
+        "YIELD_FROM",
+        "GET_YIELD_FROM_ITER",
+        "SEND",
+        "MAKE_FUNCTION",
+        # This could be supported in the future.
+        "LOAD_DEREF",
+        # Exceptions
+        "RAISE_VARARGS",
+        "SETUP_FINALLY",
+        "POP_EXCEPT",
+        # TODO: Nested loops are not supported for now.
+        # We can support them by recursively calling this translation every time
+        # for each nested loop, from the innermost loop to the outermost.
+        #
+        # E.g.
+        # loop1:
+        #   loop2:
+        #     loop3:
+        #
+        # Compile loop3, then loop2, then loop1
+        "FOR_ITER",
+        # These two are fine for the end of the loop body but not within.
+        # TODO: we can support these too. They just need to be transformed to the
+        # right RETURN_VALUEs (see code below for how to do that).
+        "JUMP_BACKWARD",
+        "JUMP_ABSOLUTE",
+        # list, set, dictionary comprehensions
+        "LIST_APPEND",
+        "SET_ADD",
+        "MAP_ADD",
+        # 3,12
+        "RETURN_CONST",
+    }
+
+    store_target: int
+
+    def __init__(self, value: "RangeIteratorVariable", **kwargs):
+        from ..source import GlobalSource
+
+        super().__init__(value, source=GlobalSource("__builtins__.range"), **kwargs)
+        self.store_target = -1
+
+    @staticmethod
+    def make_self(
+        tx,
+        value: "RangeIteratorVariable",
+        real_globals: dict,
+        # The code object the loop was defined in.
+        host_code_object: types.CodeType,
+        loop_body_instructions: List["Instruction"],
+        symbolic_locals: Dict[str, VariableTracker],
+    ):
+        if (
+            loop_items := len(value.items)
+        ) < torch._dynamo.config.for_loop_medium_size_boundary:
+            raise CannotConvertRangeToHigherOrder(
+                f"Loop of length {loop_items} too small to consider optimizing"
+            )
+        assert loop_body_instructions[-1].opname in {"JUMP_BACKWARD", "JUMP_ABSOLUTE"}
+        if loop_body_instructions[0].opname == "UNPACK_SEQUENCE":
+            raise CannotConvertRangeToHigherOrder(
+                "Unpacking a range iterator, seems wrong."
```

Comment:
Why is this wrong?  Isn't it just:
```
for a,b in ...
```
which is common

Reply:
This conversion only currently works for `range(x)`. There's no way to unpack a range iterator like that.